File: GPU-Distributed/Torrent/src/main/java/com/company/proto/handlers/SearchHandler.java
package com.company.proto.handlers;

import com.company.proto.NodeConfig;
import com.company.proto.UtilsKt;
import com.company.proto.torrent.Torrent;
import com.google.protobuf.ByteString;
import io.reactivex.schedulers.Timed;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.net.ConnectException;
import java.net.Socket;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeUnit;

import static com.company.proto.UtilsIO.readMessageFrom;
import static com.company.proto.UtilsIO.writeMessageTo;

public class SearchHandler implements Handler {
	private Torrent.Node currentNode;
	private LocalSearchHandler localSearchHandler;
	
	public SearchHandler(Map<Torrent.FileInfo, ByteString> storage, Torrent.Node currentNode) {
		this.currentNode = currentNode;
		this.localSearchHandler = new LocalSearchHandler(storage);
	}
	
	@Override
	public Torrent.Message handle(Torrent.Message message) {
		String regex = message.getSearchRequest().getRegex();
		
		if (!UtilsKt.isValidRegex(regex)) return messageError(regex);
		
		Torrent.LocalSearchRequest build = Torrent.LocalSearchRequest.newBuilder()
				.setRegex(regex)
				.build();
		
		Torrent.Message localSearchRequest = Torrent.Message.newBuilder()
				.setType(Torrent.Message.Type.LOCAL_SEARCH_REQUEST)
				.setLocalSearchRequest(build)
				.build();
		
		Torrent.Message localResponse = localSearchHandler.handle(localSearchRequest);
		Torrent.LocalSearchResponse localSearchResponse = localResponse.getLocalSearchResponse();
		Torrent.NodeSearchResult nodeSearchResult = Torrent.NodeSearchResult.newBuilder()
				.setNode(currentNode)
				.setStatus(localSearchResponse.getStatus())
				.setErrorMessage(localSearchResponse.getErrorMessage())
				.addAllFiles(localSearchResponse.getFileInfoList())
				.build();
		
		return askOtherNodesFor(nodeSearchResult, regex);
	}
	
	private Torrent.Message askOtherNodesFor(Torrent.NodeSearchResult currentNodeSearchResult, String regex) {
		Torrent.SearchResponse.Builder searchResponse = Torrent.SearchResponse.newBuilder()
				.setStatus(Torrent.Status.SUCCESS)
				.setErrorMessage("SUCCESS");
		searchResponse.addResults(currentNodeSearchResult);
		NodeConfig.INSTANCE.nodeObservable(currentNode)
//				.timeInterval(TimeUnit.MILLISECONDS)
//				.map(Timed::value)
				.subscribe(node -> {
					try (Socket socket = new Socket(node.getHost(), node.getPort())) {
						DataOutputStream output = new DataOutputStream(socket.getOutputStream());
						DataInputStream input = new DataInputStream(socket.getInputStream());
						
						Torrent.Message reqMessage = createLocalSearchRequest(regex);
						writeMessageTo(reqMessage, output);
						
						Torrent.Message resMessage = readMessageFrom(input);
						Torrent.LocalSearchResponse localSearchResponse = resMessage.getLocalSearchResponse();
						
						Torrent.NodeSearchResult nodeSearchResult = Torrent.NodeSearchResult.newBuilder()
								.setNode(node)
								.setStatus(localSearchResponse.getStatus())
								.setErrorMessage(localSearchResponse.getErrorMessage())
								.addAllFiles(localSearchResponse.getFileInfoList())
								.build();
						searchResponse.addResults(nodeSearchResult);
					} catch (IOException error) {
						Torrent.FileInfo fileInfo = Torrent.FileInfo.newBuilder().setFilename(regex).build();
						Torrent.NodeSearchResult.Builder nodeSearchResult = Torrent.NodeSearchResult.newBuilder()
								.addFiles(fileInfo)
								.setNode(node);
						if (error instanceof ConnectException) {
							nodeSearchResult.setStatus(Torrent.Status.NETWORK_ERROR);
							nodeSearchResult.setErrorMessage("NETWORK_ERROR");
						} else {
							nodeSearchResult.setStatus(Torrent.Status.PROCESSING_ERROR);
							nodeSearchResult.setErrorMessage("PROCESSING_ERROR");
						}
						searchResponse.addResults(nodeSearchResult.build());
						System.out.println("node: " + node.toString()+"\n"+error.toString());
					}
				}, Throwable::printStackTrace);
		
		return Torrent.Message.newBuilder()
				.setType(Torrent.Message.Type.SEARCH_RESPONSE)
				.setSearchResponse(searchResponse.build())
				.build();
	}
	
	private Torrent.Message createLocalSearchRequest(String regex) {
		Torrent.LocalSearchRequest localSearchRequest = Torrent.LocalSearchRequest.newBuilder()
				.setRegex(regex)
				.build();
		return Torrent.Message.newBuilder()
				.setLocalSearchRequest(localSearchRequest)
				.setType(Torrent.Message.Type.LOCAL_SEARCH_REQUEST)
				.build();
	}
	
	private Torrent.Message messageError(String filename) {
		Torrent.FileInfo fileInfo = Torrent.FileInfo.newBuilder()
				.setFilename(filename)
				.build();
		Torrent.NodeSearchResult nodeSearchResult = Torrent.NodeSearchResult
				.newBuilder()
				.setNode(currentNode)
				.setStatus(Torrent.Status.MESSAGE_ERROR)
				.setErrorMessage("MESSAGE_ERROR")
				.addFiles(fileInfo)
				.build();
		Torrent.SearchResponse searchResponse = Torrent.SearchResponse
				.newBuilder()
				.setStatus(Torrent.Status.MESSAGE_ERROR)
				.setErrorMessage("MESSAGE_ERROR")
				.addResults(nodeSearchResult)
				.build();
		return Torrent.Message.newBuilder()
				.setType(Torrent.Message.Type.SEARCH_RESPONSE)
				.setSearchResponse(searchResponse)
				.build();
	}
}

File: GPU-Distributed/Torrent/src/main/java/com/company/proto/handlers/SearchHandler.kt
package com.company.proto.handlers

import com.company.proto.*
import com.company.proto.torrent.Torrent
import com.google.protobuf.ByteString
import kotlinx.coroutines.experimental.delay

import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.IOException
import java.net.ConnectException
import java.net.Socket


class SearchHandler(
		private val storage: Map<Torrent.FileInfo, ByteString>,
		private val currentNode: Torrent.Node
) : Handler {
	private val localSearchHandler: LocalSearchHandler = LocalSearchHandler(storage)
	
	override fun handle(message: Torrent.Message): Torrent.Message {
		val regex = message.searchRequest.regex
		
		if (!regex.isValidRegex()) return messageError(regex)
		
		val localResponse = localSearchHandler.handle(localSearchRequest(regex))
		val nodeSearchResult = Torrent.NodeSearchResult.newBuilder()
				.setNode(currentNode)
				.setStatus(localResponse.localSearchResponse.status)
				.setErrorMessage(localResponse.localSearchResponse.errorMessage)
				.addAllFiles(localResponse.localSearchResponse.fileInfoList)
				.build()
		
		return askOtherNodesFor(nodeSearchResult, regex)
	}
	
	private fun localSearchRequest(regex: String): Torrent.Message {
		val build = Torrent.LocalSearchRequest.newBuilder()
				.setRegex(regex)
				.build()
		return Torrent.Message.newBuilder()
				.setType(Torrent.Message.Type.LOCAL_SEARCH_REQUEST)
				.setLocalSearchRequest(build)
				.build()
	}
	
	private fun askOtherNodesFor(currentNodeSearchResult: Torrent.NodeSearchResult, regex: String): Torrent.Message {
		println("SearchHandler - ask nodes:${currentNode.host}:${currentNode.port}")
		val searchResponse = Torrent.SearchResponse.newBuilder()
				.setStatus(Torrent.Status.SUCCESS)
				.addResults(currentNodeSearchResult)
		nodeList.filter { it != currentNode }.forEach { node ->
			try {
				Socket(node.host, node.port).use { socket ->
					val output = socket.getDataOutputStream()
					val input = socket.getDataInputStream()
					
					val reqMessage = createLocalSearchRequest(regex)
					output.writeMessage(reqMessage)
					
					val resMessage = input.readMessage()
					val localSearchResponse = resMessage.localSearchResponse
					
					val nodeSearchResult = Torrent.NodeSearchResult.newBuilder()
							.setNode(node)
							.setStatus(localSearchResponse.status)
							.setErrorMessage(localSearchResponse.errorMessage)
							.addAllFiles(localSearchResponse.fileInfoList)
							.build()
					searchResponse.addResults(nodeSearchResult)
				}
			} catch (error: IOException) {
				val fileInfo = Torrent.FileInfo.newBuilder().setFilename(regex).build()
				val nodeSearchResult = Torrent.NodeSearchResult.newBuilder()
						.addFiles(fileInfo)
						.setNode(node)
				if (error is ConnectException) {
					nodeSearchResult.status = Torrent.Status.NETWORK_ERROR
					nodeSearchResult.errorMessage = "NETWORK_ERROR"
				} else {
					nodeSearchResult.status = Torrent.Status.PROCESSING_ERROR
					nodeSearchResult.errorMessage = "PROCESSING_ERROR"
				}
				searchResponse.addResults(nodeSearchResult.build())
			}
		}
		
		return Torrent.Message.newBuilder()
				.setType(Torrent.Message.Type.SEARCH_RESPONSE)
				.setSearchResponse(searchResponse.build())
				.build()
	}
	
	private fun createLocalSearchRequest(regex: String): Torrent.Message {
		val localSearchRequest = Torrent.LocalSearchRequest.newBuilder()
				.setRegex(regex)
				.build()
		return Torrent.Message.newBuilder()
				.setLocalSearchRequest(localSearchRequest)
				.setType(Torrent.Message.Type.LOCAL_SEARCH_REQUEST)
				.build()
	}
	
	private fun messageError(filename: String): Torrent.Message {
		val fileInfo = Torrent.FileInfo.newBuilder()
				.setFilename(filename)
				.build()
		val nodeSearchResult = Torrent.NodeSearchResult
				.newBuilder()
				.setNode(currentNode)
				.setStatus(Torrent.Status.MESSAGE_ERROR)
				.setErrorMessage("Regex $filename is invalid")
				.addFiles(fileInfo)
				.build()
		val searchResponse = Torrent.SearchResponse
				.newBuilder()
				.setStatus(Torrent.Status.MESSAGE_ERROR)
				.setErrorMessage("MESSAGE_ERROR")
				.addResults(nodeSearchResult)
				.build()
		return Torrent.Message.newBuilder()
				.setType(Torrent.Message.Type.SEARCH_RESPONSE)
				.setSearchResponse(searchResponse)
				.build()
	}
}
